source/MaterialXFormat/Util.cpp

```diff
@@ -234,6 +234,97 @@ void flattenFilenames(DocumentPtr doc, const FileSearchPath& searchPath, StringR
+{
+    FileSearchPath resolvedDefinitionPath;
+    const FileSearchPath coreDefinitionPath = getCoreDefinitionPath();
+    for (FileSearchPath::ConstIterator path=coreDefinitionPath.begin(); path != coreDefinitionPath.end(); ++path)
+    {
+        resolvedDefinitionPath.append(*path);
+    }
+    const FileSearchPath environmentPath = getEnvironmentPath();
+    for (FileSearchPath::ConstIterator path=environmentPath.begin(); path != environmentPath.end(); ++path)
+    {
+        resolvedDefinitionPath.append(*path);
+    }
+    for (FileSearchPath::ConstIterator path=userDefinitionPath.begin(); path != userDefinitionPath.end(); ++path)
+    {
+        resolvedDefinitionPath.append(*path);
+    }
+    const FileSearchPath assetDefinitionPath = getAssetDefinitionPath();
+    for (FileSearchPath::ConstIterator path=assetDefinitionPath.begin(); path != assetDefinitionPath.end(); ++path)
+    {
+        resolvedDefinitionPath.append(*path);
+    }
+    if (includeSubFolders)
+    {
+        std::set<std::string> childDefinitionPathsSet;
+        FileSearchPath childDefinitionPaths;
+        for (FileSearchPath::ConstIterator path = resolvedDefinitionPath.begin(); path != resolvedDefinitionPath.end(); ++path)
+        {
+            FilePath filePath = *path;
+            FilePathVec subdirs = filePath.getSubDirectories();
+            for (FilePath subdirPath : subdirs)
+            {
+                if (childDefinitionPathsSet.count(subdirPath.asString()) == 0)
+                {
+                    childDefinitionPaths.append(subdirPath);
+                    childDefinitionPathsSet.emplace(subdirPath.asString());
+                }
+            }
+        }
+        return childDefinitionPaths;
+    }
+    else
+    {
+        return resolvedDefinitionPath;
+    }
+}
+
+FileSearchPath getResolvedTexturePath(const FileSearchPath& userTexturePath, const FileSearchPath& userDefinitionPath, bool includeSubFolders)
+{
+    FileSearchPath resolvedTexturePath;
+    const FileSearchPath assetTexturePath = getAssetDefinitionPath();
+    if (includeSubFolders)
+    {
+        FilePathVec libraryFolders;
+        for (FileSearchPath::ConstIterator path=userTexturePath.begin(); path != userTexturePath.end(); ++path)
+        {
+            libraryFolders.push_back(*path);
+        }
+        for (FileSearchPath::ConstIterator path=assetTexturePath.begin(); path != assetTexturePath.end(); ++path)
+        {
+            libraryFolders.push_back(*path);
+        }
+        FilePathVec childFolders;
+        getSubdirectories(libraryFolders, assetTexturePath, childFolders);
+        libraryFolders.insert(std::end(libraryFolders), std::begin(childFolders), std::end(childFolders));
+        for (const auto& childFolder : childFolders)
+        {
+            resolvedTexturePath.append(childFolder);
+        }
+    }
+    else
+    {
+        for (FileSearchPath::ConstIterator path=userTexturePath.begin(); path != userTexturePath.end(); ++path)
```

Comment:
Simplifying logic would be `resolvedTexturePath.append(userTexturePath);`
Same for the other one.